File: backend/internal/handler/database_admin.go
package handler

import (
	"strconv"

	"github.com/gin-gonic/gin"
	"lingosql/internal/models"
	"lingosql/internal/service"
	"lingosql/internal/utils"
)

type DatabaseAdminHandler struct {
	databaseAdminService *service.DatabaseAdminService
}

func NewDatabaseAdminHandler(databaseAdminService *service.DatabaseAdminService) *DatabaseAdminHandler {
	return &DatabaseAdminHandler{
		databaseAdminService: databaseAdminService,
	}
}

// CheckAdminPermissions 检查管理权限
func (h *DatabaseAdminHandler) CheckAdminPermissions(c *gin.Context) {
	connectionIDStr := c.Query("connection_id")
	connectionID, err := strconv.Atoi(connectionIDStr)
	if err != nil {
		utils.BadRequest(c, "无效的连接 ID")
		return
	}

	userID, ok := GetUserID(c)
	if !ok {
		return
	}
	permissions, err := h.databaseAdminService.CheckAdminPermissions(connectionID, userID)
	if err != nil {
		utils.BadRequest(c, err.Error())
		return
	}

	utils.Success(c, permissions)
}

// GetDatabaseList 获取数据库列表（带详细信息）
func (h *DatabaseAdminHandler) GetDatabaseList(c *gin.Context) {
	connectionIDStr := c.Query("connection_id")
	connectionID, err := strconv.Atoi(connectionIDStr)
	if err != nil {
		utils.BadRequest(c, "无效的连接 ID")
		return
	}

	userID, ok := GetUserID(c)
	if !ok {
		return
	}
	databases, err := h.databaseAdminService.GetDatabaseList(connectionID, userID)
	if err != nil {
		utils.BadRequest(c, err.Error())
		return
	}

	utils.Success(c, databases)
}

// GetDatabaseInfo 获取数据库详情
func (h *DatabaseAdminHandler) GetDatabaseInfo(c *gin.Context) {
	connectionIDStr := c.Query("connection_id")
	connectionID, err := strconv.Atoi(connectionIDStr)
	if err != nil {
		utils.BadRequest(c, "无效的连接 ID")
		return
	}

	databaseName := c.Param("name")
	if databaseName == "" {
		utils.BadRequest(c, "数据库名称不能为空")
		return
	}

	userID, ok := GetUserID(c)
	if !ok {
		return
	}
	info, err := h.databaseAdminService.GetDatabaseInfo(connectionID, userID, databaseName)
	if err != nil {
		utils.BadRequest(c, err.Error())
		return
	}

	utils.Success(c, info)
}

// CreateDatabase 创建数据库
func (h *DatabaseAdminHandler) CreateDatabase(c *gin.Context) {
	connectionIDStr := c.Query("connection_id")
	connectionID, err := strconv.Atoi(connectionIDStr)
	if err != nil {
		utils.BadRequest(c, "无效的连接 ID")
		return
	}

	var req models.CreateDatabaseRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.BadRequest(c, "请求参数错误: "+err.Error())
		return
	}

	userID, ok := GetUserID(c)
	if !ok {
		return
	}
	if err := h.databaseAdminService.CreateDatabase(connectionID, userID, &req); err != nil {
		utils.BadRequest(c, err.Error())
		return
	}

	utils.Success(c, nil)
}

// DropDatabase 删除数据库
func (h *DatabaseAdminHandler) DropDatabase(c *gin.Context) {
	connectionIDStr := c.Query("connection_id")
	connectionID, err := strconv.Atoi(connectionIDStr)
	if err != nil {
		utils.BadRequest(c, "无效的连接 ID")
		return
	}

	databaseName := c.Param("name")
	if databaseName == "" {
		utils.BadRequest(c, "数据库名称不能为空")
		return
	}

	userID, ok := GetUserID(c)
	if !ok {
		return
	}
	if err := h.databaseAdminService.DropDatabase(connectionID, userID, databaseName); err != nil {
		utils.BadRequest(c, err.Error())
		return
	}

	utils.Success(c, nil)
}

// RenameDatabase 重命名数据库
func (h *DatabaseAdminHandler) RenameDatabase(c *gin.Context) {
	connectionIDStr := c.Query("connection_id")
	connectionID, err := strconv.Atoi(connectionIDStr)
	if err != nil {
		utils.BadRequest(c, "无效的连接 ID")
		return
	}

	databaseName := c.Param("name")
	if databaseName == "" {
		utils.BadRequest(c, "数据库名称不能为空")
		return
	}

	var req models.RenameDatabaseRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.BadRequest(c, "请求参数错误: "+err.Error())
		return
	}

	userID, ok := GetUserID(c)
	if !ok {
		return
	}
	if err := h.databaseAdminService.RenameDatabase(connectionID, userID, databaseName, &req); err != nil {
		utils.BadRequest(c, err.Error())
		return
	}

	utils.Success(c, nil)
}

File: frontend/src/api/databaseAdmin.ts
/**
 * 数据库管理 API
 * 注意：这些API需要后端实现，当前为占位实现
 */

import request from '@/utils/request'
import type { ApiResponse } from '@/types/api'
import type { AdminPermission, DatabaseInfo, CreateDatabaseRequest, RenameDatabaseRequest } from '@/types/databaseAdmin'

/**
 * 检查当前连接用户的管理权限
 */
export async function checkAdminPermissions(connectionId: number): Promise<AdminPermission> {
  const response = await request.get<ApiResponse<AdminPermission>>('/admin/permissions/check', {
    params: { connection_id: connectionId }
  })
  return response.data
}

/**
 * 获取数据库详细信息列表
 */
export async function getDatabaseList(connectionId: number): Promise<DatabaseInfo[]> {
  const response = await request.get<ApiResponse<DatabaseInfo[]>>('/admin/databases', {
    params: { connection_id: connectionId }
  })
  return response.data
}

/**
 * 获取单个数据库详情
 */
export async function getDatabaseInfo(connectionId: number, databaseName: string): Promise<DatabaseInfo> {
  const response = await request.get<ApiResponse<DatabaseInfo>>(`/admin/databases/${databaseName}/info`, {
    params: { connection_id: connectionId }
  })
  return response.data
}

/**
 * 创建数据库
 */
export async function createDatabase(connectionId: number, data: CreateDatabaseRequest): Promise<void> {
  await request.post<ApiResponse<void>>('/admin/databases', data, {
    params: { connection_id: connectionId }
  })
}

/**
 * 删除数据库
 */
export async function deleteDatabase(connectionId: number, databaseName: string): Promise<void> {
  await request.delete<ApiResponse<void>>(`/admin/databases/${databaseName}`, {
    params: { connection_id: connectionId }
  })
}

/**
 * 重命名数据库（仅PostgreSQL）
 */
export async function renameDatabase(connectionId: number, databaseName: string, data: RenameDatabaseRequest): Promise<void> {
  await request.post<ApiResponse<void>>(`/admin/databases/${databaseName}/rename`, data, {
    params: { connection_id: connectionId }
  })
}
